ait/core/server/stream.py

```diff
@@ -101,13 +105,146 @@ def valid_workflow(self):
         return True
 
 
-class PortInputStream(Stream, PortInputClient):
+def output_stream_factory(name, inputs, outputs, handlers, zmq_args=None):
+    """
+    This factory preempts the creating of output streams directly.  It accepts
+    the same args as any given stream class and then based primarily on the
+    values in 'outputs' decides on the appropriate stream to instantiate and
+    then returns it.
+    """
+
+    parsed_output = outputs
+    if type(parsed_output) is list and len(parsed_output) > 0:
+        if len(parsed_output) > 1:
+            ait.core.log.warn(f"Additional output args discarded {parsed_output[1:]}")
+        parsed_output = parsed_output[0]
+    if type(parsed_output) is int:
+        if ait.MIN_PORT <= parsed_output <= ait.MAX_PORT:
+            return UDPOutputStream(
+                name, inputs, parsed_output, handlers, zmq_args=zmq_args
+            )
+        else:
+            raise ValueError(f"Output stream specification invalid: {outputs}")
+
+    elif type(parsed_output) is str and is_valid_address_spec(parsed_output):
+        protocol, hostname, port = parsed_output.split(":")
+        if protocol.lower() == "udp" and ait.MIN_PORT <= int(port) <= ait.MAX_PORT:
+            return UDPOutputStream(
+                name, inputs, parsed_output, handlers, zmq_args=zmq_args
+            )
+        elif protocol.lower() == "tcp" and ait.MIN_PORT <= int(port) <= ait.MAX_PORT:
+            return TCPOutputStream(
+                name, inputs, parsed_output, handlers, zmq_args=zmq_args
+            )
+        else:
+            raise ValueError(f"Output stream specification invalid: {outputs}")
+    elif parsed_output is None or (
+        type(parsed_output) is list and len(parsed_output) == 0
+    ):
+        return ZMQStream(
+            name,
+            inputs,
+            handlers,
+            zmq_args=zmq_args,
+        )
+    else:
+        raise ValueError(f"Output stream specification invalid: {outputs}")
+
+
+def input_stream_factory(name, inputs, handlers, zmq_args=None):
+    """
+    This factory preempts the creating of input streams directly.  It accepts
+    the same args as any given stream class and then based primarily on the
+    values in 'inputs' decides on the appropriate stream to instantiate and
+    then returns it.
+    """
+
+    stream = None
+    parsed_inputs = inputs
+    if type(parsed_inputs) is int:
+        parsed_inputs = [parsed_inputs]
+    if type(parsed_inputs) is str:
+        parsed_inputs = [parsed_inputs]
+
+    if type(parsed_inputs) is not list or (
+        type(parsed_inputs) is list and len(parsed_inputs) == 0
+    ):
+        raise ValueError(f"Input stream specification invalid: {parsed_inputs}")
+
+    # backwards compatability with original UDP server spec
```

Comment:
More a question than request, the output factory warns when discarding excess outputs.  Should a similar warning be included here for the non-ZMQStream cases when only first input of many is used?